src/Data/Store.elm

```diff
@@ -0,0 +1,211 @@
+module Data.Store exposing
+    ( AudienceFolderID
+    , AudienceID
+    , AudienceLevel
+    , Selection(..)
+    , Selector(..)
+    , Store
+    , getFolderByID
+    , init
+    , select
+    )
+
+import Data.Audience as Audience exposing (Audience, AudienceType)
+import Data.AudienceFolder exposing (AudienceFolder)
+import Dict exposing (Dict)
+
+
+type alias AudienceFolderID =
+    Int
+
+
+type alias AudienceID =
+    Int
+
+
+type alias AudienceLevel =
+    { folders : List AudienceFolder
+    , audiences : List Audience
+    }
+
+
+type alias Level =
+    { folders : List AudienceFolderID
+    , curated : List AudienceID
+    , authored : List AudienceID
+    }
+
+
+emptyLevel : Level
+emptyLevel =
+    Level [] [] []
+
+
+insertFolderToLevel : AudienceFolderID -> Level -> Level
+insertFolderToLevel folderID level =
+    { level | folders = folderID :: level.folders }
+
+
+insertAudienceToLevel : Audience -> List AudienceID -> Level -> ( List AudienceID, Level )
+insertAudienceToLevel audience shared level =
+    case audience.type_ of
+        Audience.Authored ->
+            ( shared, { level | authored = audience.id :: level.authored } )
+
+        Audience.Shared ->
+            ( audience.id :: shared, level )
+
+        Audience.Curated ->
+            ( shared, { level | curated = audience.id :: level.curated } )
+
+
+type alias Levels =
+    Dict AudienceFolderID Level
+
+
+type Store
+    = Store
+        { folders : Dict AudienceFolderID AudienceFolder
+        , audiences : Dict AudienceID Audience
+        }
+        { root : Level
+        , levels : Levels
+        , shared : List AudienceID
+        }
+
+
+init : List AudienceFolder -> List Audience -> Store
+init listOfFolders listOfAudiences =
+    let
+        data0 =
+            { folders = Dict.empty
+            , audiences = Dict.empty
+            }
+
+        relations0 =
+            { root = emptyLevel
+            , levels = Dict.fromList (List.map (\folder -> ( folder.id, emptyLevel )) listOfFolders)
+            , shared = []
+            }
+
+        ( data1, relations1 ) =
```

Comment:
I'd prefer these `(data1, relations1)` and `(data2, relations2)` things to use some extracted function whose name tells us better what actually happened (ie. something like `processFolders` and `processAudiences`)

Reply:
I've dave a little bit deeper and remembered
why I'm using the composite (`case (msg, model of`) approach.

Let's say we have `Model` and `Msg`.
Each of the variants must be handled according postfix (`ModelA` with `MsgA`, etc...).

```elm
type Model
    = ModelA Int
    | ModelB Int
    | ModelC Int


type Msg
    = MsgA Int
    | MsgB Int
    | MsgC Int
```

The composite function will look like this:

```elm
update : Msg -> Model -> Model
update msg model =
    case ( msg, model ) of
        ( MsgA a, ModelA x ) ->
            ModelB (a + x)

        ( MsgA _, _ ) ->
            model

        ( MsgB b, ModelB x ) ->
            ModelC (b * x)

        ( MsgB _, _ ) ->
            model

        ( MsgC c, ModelC x ) ->
            ModelA (c - x)

        ( MsgC _, _ ) ->
            model
```

It prevents you from not handling new variants.
Let's say we add `MsgAA` to handle when `ModelA`:


```elm
type Msg
    = MsgA Int
    | MsgAA
    | MsgB Int
    | MsgC Int

-- compiling error
```

Compiler will say us to handle the case:

```elm
update : Msg -> Model -> Model
update msg model =
    case ( msg, model ) of
        ( MsgA a, ModelA x ) ->
            ModelB (a + x)

        ( MsgA _, _ ) ->
            model

        ( MsgAA, ModelA x ) ->
            ModelB x

        ( MsgAA _, _ ) ->
            model

        ( MsgB b, ModelB x ) ->
            ModelC (b * x)

        ( MsgB _, _ ) ->
            model

        ( MsgC c, ModelC x ) ->
            ModelA (c - x)

        ( MsgC _, _ ) ->
            model
```

But let's separate the `update` to multiple functions:

```elm
updateA : Msg -> Int -> Model
updateA msg x =
    case msg of
        MsgA a ->
            ModelB (a + x)

        _ ->
            ModelA x


updateB : Msg -> Int -> Model
updateB msg x =
    case msg of
        MsgB b ->
            ModelC (b * x)

        _ ->
            ModelB x


updateC : Msg -> Int -> Model
updateC msg x =
    case msg of
        MsgC c ->
            ModelA (c - x)

        _ ->
            ModelC x
```


The problem is that when I add one more Msg for ModelA
the compiler will say me nothing but the case will not be handled.

```elm
type Msg
    = MsgA Int
    | MsgAA
    | MsgB Int
    | MsgC Int

-- no more changes but still compiling
```

To prevent this situation I could handle all of the cases dirrectly:

```elm
updateA : Msg -> Int -> Model
updateA msg x =
    case msg of
        MsgA a ->
            ModelB (a + x)

        MsgB _ ->
            ModelA x

        MsgC _ ->
            ModelA x


updateB : Msg -> Int -> Model
updateB msg x =
    case msg of
        MsgA _ ->
            ModelB x

        MsgB b ->
            ModelC (b * x)

        MsgC _ ->
            ModelB x


updateC : Msg -> Int -> Model
updateC msg x =
    case msg of
        MsgA _ ->
            ModelC x

        MsgB _ ->
            ModelC x

        MsgC c ->
            ModelA (c - x)
```

As you can see the previous problem is solved, but another created:
you mast describe too much.

Otherwise composite approach prevent you from both
of the issues. `update` is becoming bigger and bigger but instead of
splitting by model you could split by cases, like:


```elm
hugeACase : Int -> Int -> Model
hugeACase a x =
    ModelB (a + x)


hugeBCase : Int -> Int -> Model
hugeBCase b x =
    ModelC (b * x)


hugeCCase : Int -> Int -> Model
hugeCCase c x =
    ModelC (c - x)


update : Msg -> Model -> Model
update msg model =
    case ( msg, model ) of
        ( MsgA a, ModelA x ) ->
            hugeACase a x

        ( MsgA _, _ ) ->
            model

        ( MsgB b, ModelB x ) ->
            hugeBCase b x

        ( MsgB _, _ ) ->
            model

        ( MsgC c, ModelC x ) ->
            hugeCCase

        ( MsgC _, _ ) ->
            model
```